pkg/procstats/procstats.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2019 Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package procstats
+
+import (
+	"io/ioutil"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/intel/cri-resource-manager/pkg/log"
+	"github.com/intel/cri-resource-manager/pkg/sysfs"
+)
+
+// CPUTimeStat is used to calculate the CPU usage.
+type CPUTimeStat struct {
+	sync.RWMutex
+	PrevIdleTime       []uint64
+	PrevTotalTime      []uint64
+	CurIdleTime        []uint64
+	CurTotalTime       []uint64
+	DeltaIdleTime      []uint64
+	DeltaTotalTime     []uint64
+	CPUUsage           []float64
+	IsGetCPUUsageBegin bool
+}
+
+var (
+	// procRoot is the mount point for the proc filesystem
+	procRoot = "/proc"
+	procStat = procRoot + "/stat"
+)
+
+// GetCPUTimeStat calculates CPU usage by using the CPU time statistics from /proc/stat
+func (t *CPUTimeStat) GetCPUTimeStat() error {
+	lines, err := readLines(procStat)
+	if err != nil {
+		return err
+	}
+	// /proc/stat looks like this:
+	// cpuid: user, nice, system, idle, iowait, irq, softirq
+	// cpu  130216 19944 162525 1491240 3784 24749 17773 0 0 0
+	// cpu0 40321 11452 49784 403099 2615 6076 6748 0 0 0
+	// cpu1 26585 2425 36639 151166 404 2533 3541 0 0 0
+	// ...
+	t.Lock()
+	defer t.Unlock()
+	sys, err := sysfs.DiscoverSystem()
+	if err != nil {
+		return err
+	}
+	cpuCount := len(sys.CPUIDs())
+	for index, line := range lines {
+		if index > cpuCount {
+			break
+		}
+		split := strings.Split(line, " ")
+		if matched, _ := regexp.MatchString("cpu+\\d", split[0]); matched {
```

Comment:
nitpick: IMO it's quite an overkill to look for per-CPU entries with regexp in `/proc/stat`, especially this way which basically recompiles the exact same regexp pattern to a new internal representation on every iteration of the loop. `/proc/stat` is supposed to be part of the stable Linux kernel/userspace interface so we know what entries it will have.

The only entry starting with `"cpu"` which we want to ignore is the cumulated `"cpu ..."` one, so for instance something like this should work as well:
```
    if strings.HasPrefix(split[0], "cpu") && split[0] != "cpu" {
        ...
    }
```